File: typescript/src/common/retry.ts
export const withRetries = async <T>(
  operation: () => Promise<T> | T,
  maxAttempts: number,
  baseBackoffMs: number,
  isRetryable: (error: unknown) => boolean
): Promise<T> => {
  let lastError: unknown;
  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    try {
      return await operation();
    } catch (error) {
      lastError = error;
      if (!isRetryable(error) || attempt === maxAttempts) {
        break;
      }
      const sleepMs = baseBackoffMs * 2 ** (attempt - 1);
      const jitter = Math.floor(Math.random() * baseBackoffMs * 0.2);
      await new Promise((resolve) => setTimeout(resolve, sleepMs + jitter));
    }
  }
  throw lastError;
};

File: python/app/retry.py
import random
import time
from typing import Callable, Iterable, Type, TypeVar

T = TypeVar("T")


def with_retries(
    operation: Callable[[], T],
    retryable_exceptions: Iterable[Type[BaseException]],
    max_attempts: int,
    base_backoff_ms: int,
) -> T:
    for attempt in range(1, max_attempts + 1):
        try:
            return operation()
        except tuple(retryable_exceptions):
            if attempt == max_attempts:
                raise
            sleep_ms = base_backoff_ms * (2 ** (attempt - 1))
            jitter = random.randint(0, int(base_backoff_ms * 0.2))
            time.sleep((sleep_ms + jitter) / 1000.0)
    raise RuntimeError("retry attempts exhausted")
